### ISSUE TITLE
Multi-arch images?

### ISSUE THREAD

**Issue by abhilesh** [bbilly1/tubearchivist#6]
Thanks for working on this, it is a promising project!

I was trying to spin-up tubearchivist on a RPi 4 running 64-bit Raspbian OS and while the containers get built fine, I run into the ```standard_init_linux.go:228: exec user process caused: exec format error``` for the ```redis JSON``` and the ```tubearchivist``` containers (the Elasticsearch seems to install fine).

Is there a plan for building multi-arch images especially for arm64?

**Comment by bbilly1** [comment]
As far as I can tell, RedisJSON doesn't offer an ARM image. You might be able to build your own from the official Redis ARM image and install the JSON plugin. 

for Tube Archivist, that's just based on the latest python version, that should all translate without issues. The image would have to be rebuild specifically for that, but that's very feasible.

Honestly I don't depend very heavily on the JSON functionality of Redis, but then again, maybe a Raspberry Pi is not the ideal device to run your big Youtube index from...

Honestly, this is all very early stages, so I can't really give you a definitive answer. If you are interested, let me know if you can get RedisJSON to work on your Raspberry Pi, that would be very interesting!

**Comment by lamusmaser** [comment]
It should be noted that there are specific limitations for both RedisJSON and ElasticSearch wherein they will not run on any ARM 32-bit OS. While there is a workaround option to build RedisJSON manually for ARM32, ElasticSearch versions 7.0.0+ require Java as part of the package, and there is not a 32-bit option available anymore with the minimum required version. As such, ARM64 is the only available option.

Building a package for ARM64 with all prerequisites should be feasible.

**Comment by garret** [comment]
This means that an ARM64 version will also be in the work or is still not in your plans?

**Comment by lamusmaser** [comment]
I can work on determining what exactly would be required to get it working. I did have the main application of TA working with ARM32, so ARM64 should be feasible. I don't have one readily available at the moment, so I'll see if I can spin up a free-tier cloud instance somewhere to build and test. Doing that will allow me to setup an adjacent x64 VM, as well, then we can ensure testing works on both.

Once the initial testing is done, then it could probably be integrated into the testing/build cycle without much additional effort.

**Comment by abhilesh** [comment]
I've been trying to build docker images for ```arm64``` but keep running into issues while building the RedisJSON image. These are ```llvm``` related issues as docker cannot find packages for ```clang10``` in the debian repositories. Work got hectic so I hadn't had much time to troubleshoot it yet, but maybe we can collaborate on this one.

**Comment by lamusmaser** [comment]
I was able to setup my multiarch builder yesterday, so I will let you know when I have the full build sequence completed. I will let you know once we have confirmed multiarch images that work for rejson and elasticsearch. If you get to it first, please reference here. After we have those, then we will ensure that the base tubearchivist application is built with the associated architecture.

**Comment by thebiblelover7** [comment]
Hopefully this will be done soon, I really want to set this up on my RPi!

**Comment by MaciejR89** [comment]
It would be great if you could make this work in a Raspberry Pi.
It's time to change YoutubeDL-Material for something better, and this project looks promising.

**Comment by lamusmaser** [comment]
# Instructions for preparing environment for multiarch builds [Tested on Ubuntu-amd64]

## Reading on buildx
[Docker Buildx - Building with Buildx](https://docs.docker.com/buildx/working-with-buildx/)
[Docker Blog - Building Multi-arch Images the Easy Way](https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/)
[Docker Blog - Multiplatform Docker Builds](https://www.docker.com/blog/multi-platform-docker-builds/)

## Tested Procedures for Build Environment

##### Note: There are procedures in the Docker Buildx link that differ slightly. These utilize docker images to perform the installations, which can be helpful for untested variants.

### Install Docker and associated packages
`sudo apt-get update`
`sudo apt-get remove docker docker-engine docker.io containerd runc`
`sudo apt-get install -y apt-transport-https ca-certificates curl gnupg lsb-release`
`sudo apt-get install -y binfmt-support qemu-user-static`
`curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg`
`echo "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null`
`sudo apt-get update`
`sudo apt-get install -y docker-ce docker-ce-cli containerd.io docker-compose`

### Setup qemu multiarch environment locally
`docker run --rm --privileged multiarch/qemu-user-static --reset -p yes`

### Check buildx available architectures
`docker buildx ls`
##### Should see linux/amd64 and linux/arm64 present in the list of available architectures

### Create docker build instance
`docker buildx create --name tubearchivist`
`docker buildx use tubearchivist`
`docker buildx inspect --bootstrap`

### Create test Dockerfile
#### Note: This should not be used in the same directory as an existing `Dockerfile`, as this will replace the current `Dockerfile`.
`echo FROM ubuntu:20.04 > Dockerfile`
`echo RUN apt-get update && apt-get install -y curl >> Dockerfile`
`docker buildx build --platform linux/amd64,linux/arm64 -t <username>/multiarch-demo:latest --push .`
##### If you don't want to push this image, you can instead just create it locally:
`docker buildx build --platform linux/amd64,linux/arm64 -t <username>/multiarch-demo:latest .`

## Confirm that the images created have the architecture platforms present.
`docker buildx ls`
Output:
>NAME/NODE        DRIVER/ENDPOINT             STATUS  PLATFORMS
>tubearchivist *  docker-container
  tubearchivist0 unix:///var/run/docker.sock running `linux/amd64`, `linux/arm64`, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6
>default          docker
  default        default                     running `linux/amd64`, `linux/arm64`, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/arm/v7, linux/arm/v6

`docker manifest inspect --verbose <username>/multiarch-demo:latest`
Output:
>ubuntu@ubuntu-dev-amd64:~/tubearchivist-dev/tubearchivist$ docker manifest inspect --verbose lamusmaser/multiarch-demo
[        {
                "Ref": "docker.io/lamusmaser/multiarch-demo:latest@sha256:70388b67380f52b866d93eb01f18d724ea7a82416edea4d5da2b114f333d49ed",
                "Descriptor": {
                        "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                        "digest": "sha256:70388b67380f52b866d93eb01f18d724ea7a82416edea4d5da2b114f333d49ed",
                        "size": 740,
                        "platform": {
                                `"architecture": "amd64"`,
                                "os": "linux"
                        }                },
                "SchemaV2Manifest": {
                        "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                        "schemaVersion": 2,
                        "config": {
                                "mediaType": "application/vnd.docker.container.image.v1+json",
                                "digest": "sha256:7e65933108767b24db311c9af29def7fcce490b9d972d654da664b253e024865",
                                "size": 841
                        },
                        "layers": [                                {
                                        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
                                        "digest": "sha256:f3ef4ff62e0da0ef761ec1c8a578f3035bef51043e53ae1b13a20b3e03726d17",
                                        "size": 28568914
                                },                                {
                                        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
                                        "digest": "sha256:b17092b76c5a8074617eabe82fbbec9c7dde7f86f81a15c34a4f609461e720cc",
                                        "size": 28995462
                                }                        ]
                }        },        {
                "Ref": "docker.io/lamusmaser/multiarch-demo:latest@sha256:f76dd4005537609679ae2b8202ff01cf3d2f3b7a286fa465b92c562779074731",
                "Descriptor": {
                        "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                        "digest": "sha256:f76dd4005537609679ae2b8202ff01cf3d2f3b7a286fa465b92c562779074731",
                        "size": 740,
                        "platform": {
                                `"architecture": "arm64"`,
                                "os": "linux"
                        }                },
                "SchemaV2Manifest": {
                        "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
                        "schemaVersion": 2,
                        "config": {
                                "mediaType": "application/vnd.docker.container.image.v1+json",
                                "digest": "sha256:82be38de0779f395aa964c0f4eef31fad233d667ade008377015bf8c727059b2",
                                "size": 841
                        },
                        "layers": [                                {
                                        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
                                        "digest": "sha256:a2e448ef5a4dfc8e290db319d98910aa96a3abfcf38ae90bbac21672b8438d9e",
                                        "size": 27172405
                                },                                {
                                        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
                                        "digest": "sha256:e905af203e22c7dda556aeef1e2d085d62cc813977811a73010409ec9b4a899a",
                                        "size": 26086051
                                }                        ]
                }        } ]

## Next steps
A PR will be created to build multiarch by default using [deploy.sh](https://github.com/bbilly1/tubearchivist/blob/3500044b0d947d36c1bfdc7945124e21b92b4ead/deploy.sh).

**Comment by lamusmaser** [comment]
> Were you able to build an `arm64` image for RedisJSON too @lamusmaser?

I have not worked on building that, yet. I will get to it next.

**Comment by abhilesh** [comment]
Thanks a lot for the hard work, @lamusmaser & @bbilly1. I'll spin up the containers and test it out!

**Comment by sdanfa** [comment]
HI,I'm waiting to raspberry pi 4 64bit compatibility too. Please let me know once you got arm64 images working for docker.

**Comment by bbilly1** [comment]
@abhilesh @sdanfa @lamusmaser Do you have any insights on this one? Is this working on your devices?

Any feedback would be much appreciated!

**Comment by abhilesh** [comment]
@bbilly1 Can confirm install works fine on RPi 4 running Ubuntu Server 20.04 (switched OS since the last time). 
I used the example `docker-compose.yml` file with the `bbilly1/rejson` image for RedisJSON.

I will test out the performance and report back.

**Comment by bbilly1** [comment]
OK, I'm closing this issue here for now, confident enough now to remove the "looking for feedback" note in the readme.
If you encounter any problems coming up, feel free to open a separate issue. 

Thank you for everybody contributing, great teamwork! :)
